### PR TITLE
feat: allow embedding of Instagram posts on normal pages

### LINKED ISSUE #893: [Bug] Cannot embed IG post via CMS
### Problem

Users are unable to embed Instagram posts by copying the content from Instagram's embed instructions and pasting it into the CMS.

Screenshot 2022-05-24 at 5 39 57 PM

### Details
Instagram provides an embed snippet as follows:

```
<blockquote class="instagram-media" data-instgrm-permalink="https://www.instagram.com/p/Cduu7D9JX_M/?utm_source=ig_embed&amp;utm_campaign=loading" data-instgrm-version="14" style=" background:#FFF; border:0; border-radius:3px; box-shadow:0 0 1px 0 rgba(0,0,0,0.5),0 1px 10px 0 rgba(0,0,0,0.15); margin: 1px; max-width:540px; min-width:326px; padding:0; width:99.375%; width:-webkit-calc(100% - 2px); width:calc(100% - 2px);"><div style="padding:16px;"> <a href="https://www.instagram.com/p/Cduu7D9JX_M/?utm_source=ig_embed&amp;utm_campaign=loading" style=" background:#FFFFFF; line-height:0; padding:0 0; text-align:center; text-decoration:none; width:100%;" target="_blank"> <div style=" display: flex; flex-direction: row; align-items: center;"> <div style="background-color: #F4F4F4; border-radius: 50%; flex-grow: 0; height: 40px; margin-right: 14px; width: 40px;"></div> <div style="display: flex; flex-direction: column; flex-grow: 1; justify-content: center;"> <div style=" background-color: #F4F4F4; border-radius: 4px; flex-grow: 0; height: 14px; margin-bottom: 6px; width: 100px;"></div> <div style=" background-color: #F4F4F4; border-radius: 4px; flex-grow: 0; height: 14px; width: 60px;"></div></div></div><div style="padding: 19% 0;"></div> <div style="display:block; height:50px; margin:0 auto 12px; width:50px;"><svg width="50px" height="50px" viewBox="0 0 60 60" version="1.1" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink"><g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd"><g transform="translate(-511.000000, -20.000000)" fill="#000000"><g><path d="M556.869,30.41 C554.814,30.41 553.148,32.076 553.148,34.131 C553.148,36.186 554.814,37.852 556.869,37.852 C558.924,37.852 560.59,36.186 560.59,34.131 C560.59,32.076 558.924,30.41 556.869,30.41 M541,60.657 C535.114,60.657 530.342,55.887 530.342,50 C530.342,44.114 535.114,39.342 541,39.342 C546.887,39.342 551.658,44.114 551.658,50 C551.658,55.887 546.887,60.657 541,60.657 M541,33.886 C532.1,33.886 524.886,41.1 524.886,50 C524.886,58.899 532.1,66.113 541,66.113 C549.9,66.113 557.115,58.899 557.115,50 C557.115,41.1 549.9,33.886 541,33.886 M565.378,62.101 C565.244,65.022 564.756,66.606 564.346,67.663 C563.803,69.06 563.154,70.057 562.106,71.106 C561.058,72.155 560.06,72.803 558.662,73.347 C557.607,73.757 556.021,74.244 553.102,74.378 C549.944,74.521 548.997,74.552 541,74.552 C533.003,74.552 532.056,74.521 528.898,74.378 C525.979,74.244 524.393,73.757 523.338,73.347 C521.94,72.803 520.942,72.155 519.894,71.106 C518.846,70.057 518.197,69.06 517.654,67.663 C517.244,66.606 516.755,65.022 516.623,62.101 C516.479,58.943 516.448,57.996 516.448,50 C516.448,42.003 516.479,41.056 516.623,37.899 C516.755,34.978 517.244,33.391 517.654,32.338 C518.197,30.938 518.846,29.942 519.894,28.894 C520.942,27.846 521.94,27.196 523.338,26.654 C524.393,26.244 525.979,25.756 528.898,25.623 C532.057,25.479 533.004,25.448 541,25.448 C548.997,25.448 549.943,25.479 553.102,25.623 C556.021,25.756 557.607,26.244 558.662,26.654 C560.06,27.196 561.058,27.846 562.106,28.894 C563.154,29.942 563.803,30.938 564.346,32.338 C564.756,33.391 565.244,34.978 565.378,37.899 C565.522,41.056 565.552,42.003 565.552,50 C565.552,57.996 565.522,58.943 565.378,62.101 M570.82,37.631 C570.674,34.438 570.167,32.258 569.425,30.349 C568.659,28.377 567.633,26.702 565.965,25.035 C564.297,23.368 562.623,22.342 560.652,21.575 C558.743,20.834 556.562,20.326 553.369,20.18 C550.169,20.033 549.148,20 541,20 C532.853,20 531.831,20.033 528.631,20.18 C525.438,20.326 523.257,20.834 521.349,21.575 C519.376,22.342 517.703,23.368 516.035,25.035 C514.368,26.702 513.342,28.377 512.574,30.349 C511.834,32.258 511.326,34.438 511.181,37.631 C511.035,40.831 511,41.851 511,50 C511,58.147 511.035,59.17 511.181,62.369 C511.326,65.562 511.834,67.743 512.574,69.651 C513.342,71.625 514.368,73.296 516.035,74.965 C517.703,76.634 519.376,77.658 521.349,78.425 C523.257,79.167 525.438,79.673 528.631,79.82 C531.831,79.965 532.853,80.001 541,80.001 C549.148,80.001 550.169,79.965 553.369,79.82 C556.562,79.673 558.743,79.167 560.652,78.425 C562.623,77.658 564.297,76.634 565.965,74.965 C567.633,73.296 568.659,71.625 569.425,69.651 C570.167,67.743 570.674,65.562 570.82,62.369 C570.966,59.17 571,58.147 571,50 C571,41.851 570.966,40.831 570.82,37.631"></path></g></g></g></svg></div><div style="padding-top: 8px;"> <div style=" color:#3897f0; font-family:Arial,sans-serif; font-size:14px; font-style:normal; font-weight:550; line-height:18px;">View this post on Instagram</div></div><div style="padding: 12.5% 0;"></div> <div style="display: flex; flex-direction: row; margin-bottom: 14px; align-items: center;"><div> <div style="background-color: #F4F4F4; border-radius: 50%; height: 12.5px; width: 12.5px; transform: translateX(0px) translateY(7px);"></div> <div style="background-color: #F4F4F4; height: 12.5px; transform: rotate(-45deg) translateX(3px) translateY(1px); width: 12.5px; flex-grow: 0; margin-right: 14px; margin-left: 2px;"></div> <div style="background-color: #F4F4F4; border-radius: 50%; height: 12.5px; width: 12.5px; transform: translateX(9px) translateY(-18px);"></div></div><div style="margin-left: 8px;"> <div style=" background-color: #F4F4F4; border-radius: 50%; flex-grow: 0; height: 20px; width: 20px;"></div> <div style=" width: 0; height: 0; border-top: 2px solid transparent; border-left: 6px solid #f4f4f4; border-bottom: 2px solid transparent; transform: translateX(16px) translateY(-4px) rotate(30deg)"></div></div><div style="margin-left: auto;"> <div style=" width: 0px; border-top: 8px solid #F4F4F4; border-right: 8px solid transparent; transform: translateY(16px);"></div> <div style=" background-color: #F4F4F4; flex-grow: 0; height: 12px; width: 16px; transform: translateY(-4px);"></div> <div style=" width: 0; height: 0; border-top: 8px solid #F4F4F4; border-left: 8px solid transparent; transform: translateY(-4px) translateX(8px);"></div></div></div> <div style="display: flex; flex-direction: column; flex-grow: 1; justify-content: center; margin-bottom: 24px;"> <div style=" background-color: #F4F4F4; border-radius: 4px; flex-grow: 0; height: 14px; margin-bottom: 6px; width: 224px;"></div> <div style=" background-color: #F4F4F4; border-radius: 4px; flex-grow: 0; height: 14px; width: 144px;"></div></div></a><p style=" color:#c9c8cd; font-family:Arial,sans-serif; font-size:14px; line-height:17px; margin-bottom:0; margin-top:8px; overflow:hidden; padding:8px 0 7px; text-align:center; text-overflow:ellipsis; white-space:nowrap;"><a href="https://www.instagram.com/p/Cduu7D9JX_M/?utm_source=ig_embed&amp;utm_campaign=loading" style=" color:#c9c8cd; font-family:Arial,sans-serif; font-size:14px; font-style:normal; font-weight:normal; line-height:17px; text-decoration:none;" target="_blank">A post shared by SPCA Singapore Adoption (@spcasgadopt)</a></p></div></blockquote><script async src="//www.instagram.com/embed.js"></script>
```

It's quite a long snippet, but the key issue is that there is a `<script async src>` call to `www.instagram.com`.
```
<script async src="//www.instagram.com/embed.js"></script>
```

This is prevented by the `checkCSP()` function.

### Solution

We could load the Instagram embed.js file in each EditPage view by default, and get users to modify the embed instructions before pasting it into the editor.

### PR DESCRIPTION
**Pull request by dcshzj**
## Problem

Closes #893.

## Solution

**Breaking Changes**

- [ ] Yes - this PR contains breaking changes
- [x] No - this PR is backwards compatible

**Bug Fixes**:

- Allow &lt;script&gt; tags with the `src` attribute set to those hostnames allowed under the Content Security Policy to be saved.
- Added tests to verify CSP violations for &lt;script&gt; `src` attributes.

**Caveats**:

- The embedded Instagram post will appear in the Isomer CMS preview as shown in the screenshot below. However, the embed.js JavaScript file will not be loaded. Hence, the embedded post will not be interactive.
- To allow the embedded post to be interactive in the Isomer CMS preview, additional configuration will need to be made to DOMPurify. More details are available in [#893 issue comment](https://github.com/isomerpages/isomercms-frontend/issues/893#issuecomment-1211755894).

## Before & After Screenshots

**BEFORE**:

Screenshot 2022-08-18 at 11 00 58

**AFTER**:

Screenshot 2022-08-18 at 11 01 39

## Tests

**Unit tests**:
- Run `npm run tests`

**Smoke tests**:
- Run `npm run dev`.
- Log in and navigate to any site and any page.
- Add the Instagram embed code into the Markdown editor and save.
- Navigate to the generated staging site for the site that you have edited.
- Verify that the interactive embedded Instagram post is viewable on the page that you have changed.

## Deploy Notes

*None*

### CODE REVIEW

File: src/layouts/EditPage.jsx
Comment (seaerchin) [inline src/layouts/EditPage.jsx:136]: correct me if i'm wrong but one of the means of xss violation is through unvalidated `script` tags, which is what we seem to be allowing here.

should we make this more specific - ie, to only allow the sites that you've listed in #893?

**Review by seaerchin — COMMENTED** [review]
i'm not entirely sure on some of the technical details by this PR but we seem to be weakening our checks here so i'm q hesistant ._.

File: src/__tests__/utils.jsx
Comment (dcshzj) [reply src/__tests__/utils.jsx]: Good point! I added a comment before this line to explain the difference, thanks!

File: src/utils/cspUtils.js
Comment (dcshzj) [reply src/utils/cspUtils.js:27]: For `//something.com` to resolve to `http://something.com`, the user will have to be browsing the page via a `http://` page. I think that is okay because (1) I'm assuming all Isomer sites are in https and (2) the source will redirect to the https version if possible.

I added this line of code so that when the user copy-pastes the embed code inside the Markdown editor, no additional changes are necessary. The embed code comes with the source URL being `//www.instagram.com/embed.js`.

File: src/layouts/EditPage.jsx
Comment (dcshzj) [reply src/layouts/EditPage.jsx:136]: The part where only the sites listed in #893 are allowed as the source URL is handled by the `checkCSP()` function. My assumption is that if it is prevented by the `checkCSP()` function, it will not reach this segment of the code.

If there is any inline script, that would also be blocked by CSP.

Comment (dcshzj) [reply src/layouts/EditPage.jsx]: Inline scripts are blocked from loading on the actual site due to the CSP, but I updated the code to block it from saving as well.

**Comment by alexanderleegs** [comment]
> i'm not entirely sure on some of the technical details by this PR but we seem to be weakening our checks here so i'm q hesistant ._.

I agree on the security portion - I don't know enough about the security portion of this, but given that this is a deal breaker for sites currently + we're planning to do our vapt soon, I was thinking that we could put this into our identity/collaborators base branch (once that's up) for now, to be tested together with the rest of our new features, and we can release it once it's been verified to not have any security issues. Any thoughts @prestonlimlianjie?

**Review by alexanderleegs — COMMENTED** [review]
Also i think there seems to be some linting issues that are causing our ci to fail, could you run lint-fix and see if that resolves the issue?

File: src/utils/cspUtils.js
Comment (seaerchin) [reply src/utils/cspUtils.js:27]: ok so let me surmise my understanding so far just to ensure we're on the same page!

1. protocol relative URLs resolve relative to the user agent
2. isomer enforces all sites to be on **https**
3. by extension, **all** protocol relative URLs will hence resolve to **https://** as the user agent will be on `https://` whilst viewing isomer sites

**Comment by seaerchin** [comment]
so i think this PR has alot of context behind it + unstated assumptions (this part is kinda legacy HAHA); i'm not entirely sure of the underlying mechanics behind how this PR works due to the previous one but from what i gather, this PR is safe b/c of the [comment](https://github.com/isomerpages/isomercms-frontend/pull/1019#discussion_r948949500) that i posted earlier.

that being said, could we possibly document this (our CSP policy + why `DOMPurify` works even w/ replacing `https://` urls w/ PR-URLs) down in the code, so that 1. assumptions are _explicit_ not implicit. 2. confusing sections of the code (for me, this was the PR-URL section as it seemed to be weakening our security guarantee) have existing context for clarity. 

i'll approve once this is done!

**Comment by dcshzj** [comment]
@seaerchin Thanks for the feedback! I added the comments to make it a little clearer. Also, should I mark this PR as only partially resolving #893? The preview panel will still not show the interactive Instagram post, although now we allow the user save the page.

File: src/layouts/EditPage.jsx
Comment (seaerchin) [inline src/layouts/EditPage.jsx:141]: btw this is case agnostic right? the `DOMPurify` repo seemed to have lowercase for `tagNames` (see [here](https://github.com/cure53/DOMPurify/blob/main/src/tags.js) for definitions and [here](https://github.com/cure53/DOMPurify/blob/main/demos/hooks-mentaljs-demo.html#L33) for their example where it's lower cased) but if you've tested this then that's fine

**Review by seaerchin — DISMISSED** [review]
lgtm with minor clarification; hold off on merging - should still double check with @prestonlimlianjie as he's the most familiar w/ isomer infra set up but this seems fine to me

**Comment by seaerchin** [comment]
> Also, should I mark this PR as only partially resolving #893? The preview panel will still not show the interactive Instagram post, although now we allow the user save the page.

i don't have an opinion on this - both options (partial resolve and document further steps on existing PR vs resolve old issue and open new one) seem ok as long as the current limitations + future follow-ups are written somewhere

File: src/layouts/EditPage.jsx
Comment (dcshzj) [reply src/layouts/EditPage.jsx:141]: Hmm not really, in this case the `el.element.tagName` attribute will have `SCRIPT` in uppercase, as the `el.element` object is the [native Element object](https://developer.mozilla.org/en-US/docs/Web/API/Element/tagName).

> For DOM trees which represent HTML documents, the returned tag name is always in the canonical upper-case form. For example, `tagName` called on a &lt;div&gt; element returns `"DIV"`.

Comment (seaerchin) [reply src/layouts/EditPage.jsx:141]: interesting. the example given states that the dirty element is supposed to be html so i'm not too sure how this works as their example gives it as the lower-cased form ._. 

given that their definitions are in lower case, i think they might have abstracted this away from us and the casing doesn't actually matter. please resolve this if you've tested it and found it to work!

**Comment by prestonlimlianjie** [comment]
Had a quick question @dcshzj - how challenging is it to implement option 2 in this proposal list https://github.com/isomerpages/isomercms-frontend/issues/893#issuecomment-1211755894 ? It doesn't need to necessarily be a beautiful UI component for now - it could even be a jankier solution.

For example:
- All EditPage layouts have the `<script async src="https://www.instagram.com/embed.js"></script>` included by default
- All corresponding Jekyll output pages will also have the same embed.js file (not ideal, but ok)
- The user behavior is as follows
  - User goes to IG to copy embed code
  - User needs to remove the last line of the embed code
  - User then pastes the copy-pasted section into the editor

Sample embed snippet
```
<blockquote class="instagram-media" data-instgrm-captioned data-instgrm-permalink="https://www.instagram.com/p/Chi83RjpDhs/?utm_source=ig_embed&amp;utm_campaign=loading" data-instgrm-version="14" style=" background:#FFF; border:0; border-radius:3px; box-shadow:0 0 1px 0 rgba(0,0,0,0.5),0 1px 10px 0 rgba(0,0,0,0.15); margin: 1px; max-width:540px; min-width:326px; padding:0; width:99.375%; width:-webkit-calc(100% - 2px); width:calc(100% - 2px);"><div style="padding:16px;"> <a href="https://www.instagram.com/p/Chi83RjpDhs/?utm_source=ig_embed&amp;utm_campaign=loading" style=" background:#FFFFFF; line-height:0; padding:0 0; text-align:center; text-decoration:none; width:100%;" target="_blank"> <div style=" display: flex; flex-direction: row; align-items: center;"> <div style="background-color: #F4F4F4; border-radius: 50%; flex-grow: 0; height: 40px; margin-right: 14px; width: 40px;"></div> <div style="display: flex; flex-direction: column; flex-grow: 1; justify-content: center;"> <div style=" background-color: #F4F4F4; border-radius: 4px; flex-grow: 0; height: 14px; margin-bottom: 6px; width: 100px;"></div> <div style=" background-color: #F4F4F4; border-radius: 4px; flex-grow: 0; height: 14px; width: 60px;"></div></div></div><div style="padding: 19% 0;"></div> <div style="display:block; height:50px; margin:0 auto 12px; width:50px;"><svg width="50px" height="50px" viewBox="0 0 60 60" version="1.1" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink"><g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd"><g transform="translate(-511.000000, -20.000000)" fill="#000000"><g><path d="M556.869,30.41 C554.814,30.41 553.148,32.076 553.148,34.131 C553.148,36.186 554.814,37.852 556.869,37.852 C558.924,37.852 560.59,36.186 560.59,34.131 C560.59,32.076 558.924,30.41 556.869,30.41 M541,60.657 C535.114,60.657 530.342,55.887 530.342,50 C530.342,44.114 535.114,39.342 541,39.342 C546.887,39.342 551.658,44.114 551.658,50 C551.658,55.887 546.887,60.657 541,60.657 M541,33.886 C532.1,33.886 524.886,41.1 524.886,50 C524.886,58.899 532.1,66.113 541,66.113 C549.9,66.113 557.115,58.899 557.115,50 C557.115,41.1 549.9,33.886 541,33.886 M565.378,62.101 C565.244,65.022 564.756,66.606 564.346,67.663 C563.803,69.06 563.154,70.057 562.106,71.106 C561.058,72.155 560.06,72.803 558.662,73.347 C557.607,73.757 556.021,74.244 553.102,74.378 C549.944,74.521 548.997,74.552 541,74.552 C533.003,74.552 532.056,74.521 528.898,74.378 C525.979,74.244 524.393,73.757 523.338,73.347 C521.94,72.803 520.942,72.155 519.894,71.106 C518.846,70.057 518.197,69.06 517.654,67.663 C517.244,66.606 516.755,65.022 516.623,62.101 C516.479,58.943 516.448,57.996 516.448,50 C516.448,42.003 516.479,41.056 516.623,37.899 C516.755,34.978 517.244,33.391 517.654,32.338 C518.197,30.938 518.846,29.942 519.894,28.894 C520.942,27.846 521.94,27.196 523.338,26.654 C524.393,26.244 525.979,25.756 528.898,25.623 C532.057,25.479 533.004,25.448 541,25.448 C548.997,25.448 549.943,25.479 553.102,25.623 C556.021,25.756 557.607,26.244 558.662,26.654 C560.06,27.196 561.058,27.846 562.106,28.894 C563.154,29.942 563.803,30.938 564.346,32.338 C564.756,33.391 565.244,34.978 565.378,37.899 C565.522,41.056 565.552,42.003 565.552,50 C565.552,57.996 565.522,58.943 565.378,62.101 M570.82,37.631 C570.674,34.438 570.167,32.258 569.425,30.349 C568.659,28.377 567.633,26.702 565.965,25.035 C564.297,23.368 562.623,22.342 560.652,21.575 C558.743,20.834 556.562,20.326 553.369,20.18 C550.169,20.033 549.148,20 541,20 C532.853,20 531.831,20.033 528.631,20.18 C525.438,20.326 523.257,20.834 521.349,21.575 C519.376,22.342 517.703,23.368 516.035,25.035 C514.368,26.702 513.342,28.377 512.574,30.349 C511.834,32.258 511.326,34.438 511.181,37.631 C511.035,40.831 511,41.851 511,50 C511,58.147 511.035,59.17 511.181,62.369 C511.326,65.562 511.834,67.743 512.574,69.651 C513.342,71.625 514.368,73.296 516.035,74.965 C517.703,76.634 519.376,77.658 521.349,78.425 C523.257,79.167 525.438,79.673 528.631,79.82 C531.831,79.965 532.853,80.001 541,80.001 C549.148,80.001 550.169,79.965 553.369,79.82 C556.562,79.673 558.743,79.167 560.652,78.425 C562.623,77.658 564.297,76.634 565.965,74.965 C567.633,73.296 568.659,71.625 569.425,69.651 C570.167,67.743 570.674,65.562 570.82,62.369 C570.966,59.17 571,58.147 571,50 C571,41.851 570.966,40.831 570.82,37.631"></path></g></g></g></svg></div><div style="padding-top: 8px;"> <div style=" color:#3897f0; font-family:Arial,sans-serif; font-size:14px; font-style:normal; font-weight:550; line-height:18px;">View this post on Instagram</div></div><div style="padding: 12.5% 0;"></div> <div style="display: flex; flex-direction: row; margin-bottom: 14px; align-items: center;"><div> <div style="background-color: #F4F4F4; border-radius: 50%; height: 12.5px; width: 12.5px; transform: translateX(0px) translateY(7px);"></div> <div style="background-color: #F4F4F4; height: 12.5px; transform: rotate(-45deg) translateX(3px) translateY(1px); width: 12.5px; flex-grow: 0; margin-right: 14px; margin-left: 2px;"></div> <div style="background-color: #F4F4F4; border-radius: 50%; height: 12.5px; width: 12.5px; transform: translateX(9px) translateY(-18px);"></div></div><div style="margin-left: 8px;"> <div style=" background-color: #F4F4F4; border-radius: 50%; flex-grow: 0; height: 20px; width: 20px;"></div> <div style=" width: 0; height: 0; border-top: 2px solid transparent; border-left: 6px solid #f4f4f4; border-bottom: 2px solid transparent; transform: translateX(16px) translateY(-4px) rotate(30deg)"></div></div><div style="margin-left: auto;"> <div style=" width: 0px; border-top: 8px solid #F4F4F4; border-right: 8px solid transparent; transform: translateY(16px);"></div> <div style=" background-color: #F4F4F4; flex-grow: 0; height: 12px; width: 16px; transform: translateY(-4px);"></div> <div style=" width: 0; height: 0; border-top: 8px solid #F4F4F4; border-left: 8px solid transparent; transform: translateY(-4px) translateX(8px);"></div></div></div> <div style="display: flex; flex-direction: column; flex-grow: 1; justify-content: center; margin-bottom: 24px;"> <div style=" background-color: #F4F4F4; border-radius: 4px; flex-grow: 0; height: 14px; margin-bottom: 6px; width: 224px;"></div> <div style=" background-color: #F4F4F4; border-radius: 4px; flex-grow: 0; height: 14px; width: 144px;"></div></div></a><p style=" color:#c9c8cd; font-family:Arial,sans-serif; font-size:14px; line-height:17px; margin-bottom:0; margin-top:8px; overflow:hidden; padding:8px 0 7px; text-align:center; text-overflow:ellipsis; white-space:nowrap;"><a href="https://www.instagram.com/p/Chi83RjpDhs/?utm_source=ig_embed&amp;utm_campaign=loading" style=" color:#c9c8cd; font-family:Arial,sans-serif; font-size:14px; font-style:normal; font-weight:normal; line-height:17px; text-decoration:none;" target="_blank">A post shared by SPCA Singapore Adoption (@spcasgadopt)</a></p></div></blockquote> <script async src="//www.instagram.com/embed.js"></script>
```

I think the current PR is acceptable but it introduces some unnecessary security risk; I would rather that our IG integration breaks if IG makes an upstream code modification (as per the description in option 2). And we can catch this via tests.

**Comment by dcshzj** [comment]
@prestonlimlianjie I think for all 3 solutions, the same 2 issues will appear (I elaborated a little more under the "Problems to be addressed" section in https://github.com/isomerpages/isomercms-frontend/issues/893#issuecomment-1211755894):

1. The preview panel has some difficulty in showing an interactive Instagram post, especially if there are subsequent changes to the markdown that causes the preview to re-render.
2. The user needs to be able to load the Instagram embed.js script in the Isomer site.

For the proposed solution 2 specifically, I can explore this option to see if it is feasible.

> All EditPage layouts have the `<script async src="https://www.instagram.com/embed.js"></script>` included by default

This is possible, but unfortunately it does not resolve issue 1 above. This is because when the embed.js script is loaded, it modifies the `<blockquote>` HTML and changes it to display the interactive embedded post. When there are subsequent changes to the markdown text, we re-render the preview and override those changes, causing the preview to revert back to the non-interactive version of the embedded post. Based on my tests, it was not possible to get the embed.js script to reload, so I think some further explorations is necessary in this area.

> All corresponding Jekyll output pages will also have the same embed.js file (not ideal, but ok)

I think this is possible and should be trivial to implement, so we will be able to address issue 2 above.

**Comment by dcshzj** [comment]
Reverting this back to draft for now.

**Comment by dcshzj** [comment]
Updated this PR to make use of the native hooks within DOMPurify, so it should be less hacky and safer. Also added some e2e tests to verify the intended functionality.

File: src/layouts/EditPage.jsx
Comment (seaerchin) [inline src/layouts/EditPage.jsx:58]: should we just combine these two lines?

File: src/layouts/EditPage.jsx
Comment (seaerchin) [inline src/layouts/EditPage.jsx:196]: could i ask why this change was made? i think i might be lacking abit of context here

**Review by seaerchin — COMMENTED** [review]
no blockers, just wanted to get clarity on certain changes before i approve

File: src/layouts/EditPage.jsx
Comment (dcshzj) [reply src/layouts/EditPage.jsx:196]: This case assumes that the `textContent` attribute contains a non-null string. However, 2 examples where this can fail:
1. `<script></script>` which will be removed with `element` will have a null `textContent`
2. `<script async>...</script>` which will be removed with `attribute` will have a null `textContent`

Hence, for 1, we get the outerHTML which outputs the entire element that we will be removing, which also provides more context to the user on the text of the whole element that will be removed. For 2, the node name will be "async", which also lets the user know exactly the attributes that will be removed.

**Review by kishore03109 — APPROVED** [review]
lgtm

**Comment by dcshzj** [comment]
!run e2e

**Comment by dcshzj** [comment]
!run e2e

**Comment by dcshzj** [comment]
!run e2e

**Comment by dcshzj** [comment]
!run e2e

**Comment by dcshzj** [comment]
!run e2e

**Comment by dcshzj** [comment]
Merging this into `feat/identity-phase-2` as that is the branch used for VAPT.

**Pull request merged**
